### PR TITLE
Related product updates now being reflected in page previews

### PR DESCRIPTION
**Pull request by danielfmiranda**
# Description

Link to sample test page: https://foundation-s-11799-rela-3u9dyh.herokuapp.com/cms/pages/22/edit/ 
Test page credentials: admin2/admin2
Related PRs/issues: #11799

This PR introduces a dedicated template for previewing product pages, located at `previews/buyersguide/product_page.html`. 

It overrides the `{% block related_products %}` from the standard product page template to use a new template, `related_products_section.html`, which is filled with data retrieved through a new method named `preview_related_products`.

This new method leverages the existing utility `get_related_items` to fetch related product relationships directly from the CMS's in-memory state rather than the database. 

This addresses the issue where changes to related products within the CMS would not appear in the preview pane unless the page was published.

Lastly, this PR adds some tests for the new `preview_related_products` method.

# Steps to test

1. Visit https://foundation-s-11799-rela-3u9dyh.herokuapp.com/cms/pages/22/edit/ (credentials: admin2/admin2)
2. Scroll down to the related products section
3. Make a change
4. Note that the preview panel is updating accordingly
5. If everything is working as expected, testing is complete!

### CODE REVIEW

File: network-api/networkapi/wagtailpages/pagemodels/buyersguide/products.py [hunk 1]
```diff
@@ -1026,7 +1029,24 @@ def localized_related_products(self):
...
+        related_products = orderables.get_related_items(
+            self.related_product_pages.all(),
+            "related_product",
+        )
+        # FIXME: This implementation does return the localized version of each product.
+        #        But, it is inefficient. It would be better to pull all products
+        #        for the correct locale at once. This would require the above returns
+        #        a queryset of the products (rather than a list) and that we have an
+        #        efficient way of pulling all items for a given locale.
+        #        See: https://github.com/MozillaFoundation/foundation.mozilla.org/issues/9509
```
Comment (jhonatan-lopes) [inline network-api/networkapi/wagtailpages/pagemodels/buyersguide/products.py:1047]: We can change this comment for some more appropriate for this method to the tune of "we know this is inefficient but that's on purpose as we need to make sure that the preview renders the objects in memory"

File: network-api/networkapi/templates/previews/product_page.html
Comment (jhonatan-lopes) [inline network-api/networkapi/templates/previews/product_page.html:1]: Can we move this to "previews/buyersguide/product_page.html" please?

File: network-api/networkapi/templates/fragments/buyersguide/related_products_section.html
Comment (jhonatan-lopes) [inline network-api/networkapi/templates/fragments/buyersguide/related_products_section.html:1]: Awesome to have this in a separate fragment! :tada:

File: network-api/networkapi/templates/pages/buyersguide/product_page.html [hunk 2]
```diff
@@ -168,21 +168,14 @@ <h2 class="tw-h3-heading">{% trans "Tips to protect yourself" %}</h2>
...
-                    </section>
-                {% endif %}
-            {% endwith %}
+            {% block related_products %}
+                {% with related_products=product.localized_related_products %}
+                    {% if related_products %}
+                      {% include "fragments/buyersguide/related_products_section.html" with related_products=related_products %}
+                    {% endif %}
+                {% endwith %}
+            {% endblock related_products %}
```
Comment (jhonatan-lopes) [inline network-api/networkapi/templates/pages/buyersguide/product_page.html:177]: Much more readable now :tada:

**Review by jhonatan-lopes — COMMENTED** [review]
Hey @danielfmiranda, thanks for taking this work! It's all looking great! I've added some comments on where I think it can be improved

File: network-api/networkapi/wagtailpages/pagemodels/buyersguide/products.py [hunk 3]
```diff
@@ -1026,7 +1029,24 @@ def localized_related_products(self):
...
+    
+    @property
+    def preview_related_products(self) -> list:
+        """
+        Fetches related product updates for CMS page previews.
+        """
+        related_products = orderables.get_related_items(
+            self.related_product_pages.all(),
+            "related_product",
+        )
```
Comment (jhonatan-lopes) [inline network-api/networkapi/wagtailpages/pagemodels/buyersguide/products.py:1041]: I've just noticed something: this **does not** guarantee to order the related products by their `sort_order`

image

File: network-api/networkapi/wagtailpages/pagemodels/buyersguide/products.py (hunk 1 above)
Comment (danielfmiranda) [reply network-api/networkapi/wagtailpages/pagemodels/buyersguide/products.py]: Done!

File: network-api/networkapi/templates/previews/product_page.html
Comment (danielfmiranda) [reply network-api/networkapi/templates/previews/product_page.html:1]: Thanks @jhonatan-lopes! I forgot to do that. The location has now been updated :+1:

File: network-api/networkapi/utility/orderables.py [hunk 4]
```diff
@@ -1,15 +1,21 @@
...
     Return list of the related items from the queryset.
 
     For each item in the queryset, the related item stored in `related_item_field` is
     returned.
 
     """
+    # Apply ordering if needed
+    if order_by:
+        queryset = queryset.order_by(order_by)
+
```
Comment (danielfmiranda) [inline network-api/networkapi/utility/orderables.py:20]: Test

File: network-api/networkapi/wagtailpages/pagemodels/buyersguide/products.py (hunk 3 above)
Comment (danielfmiranda) [reply network-api/networkapi/wagtailpages/pagemodels/buyersguide/products.py]: Thanks for the callout @jhonatan-lopes!

[Here is a link ](https://github.com/MozillaFoundation/foundation.mozilla.org/pull/11844/files#r1483822553)to the solution I have implemented to make sure that they are always sorted by sort_order.

File: network-api/networkapi/templates/pages/buyersguide/product_page.html [hunk 5]
```diff
@@ -168,21 +168,14 @@ <h2 class="tw-h3-heading">{% trans "Tips to protect yourself" %}</h2>
...
-                {% endif %}
-            {% endwith %}
+            {% block related_products %}
+                {% with related_products=product.localized_related_products %}
+                    {% if related_products %}
+                        {% include "fragments/buyersguide/related_products_section.html" with related_products=related_products %}
+                    {% endif %}
+                {% endwith %}
+            {% endblock related_products %}
+
```
Comment (danielfmiranda) [inline network-api/networkapi/templates/pages/buyersguide/product_page.html:178]: Updated the original product page template to take advantage of the new `related_products_section` template

**Comment by danielfmiranda** [comment]
Thanks for the feedback @jhonatan-lopes! I have updated this PR accordingly and have also added some tests. This PR is now ready for re-review whenever you have a free moment 👍

File: network-api/networkapi/utility/orderables.py [hunk 6]
```diff
@@ -1,15 +1,21 @@
-from typing import TYPE_CHECKING
+from typing import TYPE_CHECKING, Optional
 
 if TYPE_CHECKING:
     from django.db import models
 
 
-def get_related_items(queryset: "models.QuerySet", related_item_field: str) -> list["models.Model"]:
+def get_related_items(
+    queryset: "models.QuerySet", related_item_field: str, order_by: Optional[str] = None
```
Comment (jhonatan-lopes) [inline network-api/networkapi/utility/orderables.py:8]: We're on Python 3.11 now, I think it's better if we use `str | None` for typing of optional paramaters

File: network-api/networkapi/wagtailpages/pagemodels/buyersguide/products.py [hunk 7]
```diff
@@ -1027,6 +1030,18 @@ def localized_related_products(self):
...
+    def preview_related_products(self) -> list:
+        """
+        Fetches related product updates for CMS page previews.
+        """
+        related_products = orderables.get_related_items(
+            self.related_product_pages.all(), "related_product", order_by="sort_order"
+        )
+        # This method, while inneficient, is neccesary as we are using it to render
+        # any related_product updates that are stored in memory, but not yet saved in the DB.
+        return [product.localized for product in related_products]
```
Comment (jhonatan-lopes) [inline network-api/networkapi/wagtailpages/pagemodels/buyersguide/products.py:1043]: Now that I'm thinking about it, the "related_products" selection only occurs in the English/default version. The other versions are synced to the default location, so they don't have a "preview" window. I think it's safe to drop the `.localized` here

File: network-api/networkapi/wagtailpages/tests/buyersguide/test_products.py [hunk 8]
```diff
@@ -121,6 +121,189 @@ def test_localized_related_products_non_default_locale(self):
...
+            page=product_page,
+            related_product=related_product_3,
+            sort_order=1,
+        )
+
+        related_preview_products = product_page.preview_related_products
+
+        self.assertEqual(len(related_preview_products), 3)
+        # Order should be product 2, product 3, product 1.
+        self.assertListEqual(
```
Comment (jhonatan-lopes) [inline network-api/networkapi/wagtailpages/tests/buyersguide/test_products.py:154]: Would it be better to use `assertCountEqual`?

File: network-api/networkapi/wagtailpages/tests/buyersguide/test_products.py [hunk 9]
```diff
@@ -121,6 +121,189 @@ def test_localized_related_products_non_default_locale(self):
...
+        # Fetch the related products again after modification
+        updated_preview_products = product_page.preview_related_products
+
+        # Updated order should now be product 1, product 2, product 3.
+        self.assertListEqual(
+            updated_preview_products,
+            [related_product_1, related_product_2, related_product_3],
+        )
+
+    def test_preview_related_products_with_non_default_locale(self):
```
Comment (jhonatan-lopes) [inline network-api/networkapi/wagtailpages/tests/buyersguide/test_products.py:259]: We might not need this anymore, see above

File: network-api/networkapi/wagtailpages/tests/buyersguide/test_products.py
Comment (jhonatan-lopes) [inline network-api/networkapi/wagtailpages/tests/buyersguide/test_products.py:1]: The tests are great but one thing I'm worried is that they're all still hitting the database. I'm thinking that we need to test the `ProductPage.serve_preview` method perharps (https://github.com/wagtail/wagtail/blob/502dd7c723a3e0740915803b8fab6b9814bb9a9d/wagtail/models/__init__.py#L796-L812).

We can then manipulate the models using the [`modelcluster`](https://github.com/wagtail/django-modelcluster) syntax directly and without saving in the database and re-check the template response to make sure that the related products section order has updated properly.

**Review by jhonatan-lopes — CHANGES_REQUESTED** [review]
Hey @danielfmiranda, thanks for this work! It's looking great, we just need to iron out some details

File: network-api/networkapi/utility/orderables.py (hunk 6 above)
Comment (danielfmiranda) [reply network-api/networkapi/utility/orderables.py]: Updated!

File: network-api/networkapi/wagtailpages/tests/buyersguide/test_products.py (hunk 8 above)
Comment (danielfmiranda) [reply network-api/networkapi/wagtailpages/tests/buyersguide/test_products.py:149]: Hey Jhonatan! I actually removed asserts relating to count in my latest commit. I will resolve this conversation for now, but can reopen/add more updates if we choose to re-implement them.

**Comment by danielfmiranda** [comment]
Hi @jhonatan-lopes! Thanks for taking a look at this PR and for the great feedback.

I was unaware that pages had a `serve_preview` method. 💡 

I ended up using the [`page.make_preview_request`](https://github.com/wagtail/wagtail/blob/502dd7c723a3e0740915803b8fab6b9814bb9a9d/wagtail/models/__init__.py#L644) method, as it does most of the request-generation work that we would need to do in order to use `serve_preview`, and then also invokes the `serve_preview` command itself.

I have also updated my tests to use `modelcluster` syntax to the best of my ability, however I will admit I had some trouble, as it looks like we use an extra "through" model for this relationship.

Can you please take a look and let me know what you think? 

Thanks again!

File: network-api/networkapi/templates/fragments/buyersguide/related_product.html [hunk 10]
```diff
@@ -13,7 +13,7 @@
             >
         </div>
         <p class="tw-body-small mt-3 mb-1">{{ related_product.company }}</p>
-        <p>{{ related_product.title }}</p>
+        <p class="product-title">{{ related_product.title }}</p>
```
Comment (danielfmiranda) [inline network-api/networkapi/templates/fragments/buyersguide/related_product.html:16]: Updated for use in tests

File: network-api/networkapi/utility/orderables.py [hunk 11]
```diff
@@ -4,12 +4,18 @@
     from django.db import models
 
 
-def get_related_items(queryset: "models.QuerySet", related_item_field: str) -> list["models.Model"]:
+def get_related_items(
+    queryset: "models.QuerySet", related_item_field: str, order_by: str | None = None
```
Comment (danielfmiranda) [inline network-api/networkapi/utility/orderables.py:8]: Updated to use str | None

**Review by jhonatan-lopes — APPROVED** [review]
LGTM :rocket: thanks @danielfmiranda

**Pull request merged**